File: code/15.三数之和.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 *
 * https://leetcode-cn.com/problems/3sum/description/
 *
 * algorithms
 * Medium (24.13%)
 * Likes:    1430
 * Dislikes: 0
 * Total Accepted:    104.8K
 * Total Submissions: 434.4K
 * Testcase Example:  '[-1,0,1,2,-1,-4]'
 *
 * 给定一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0
 * ？找出所有满足条件且不重复的三元组。
 * 
 * 注意：答案中不可以包含重复的三元组。
 * 
 * 例如, 给定数组 nums = [-1, 0, 1, 2, -1, -4]，
 * 
 * 满足要求的三元组集合为：
 * [
 * ⁠ [-1, 0, 1],
 * ⁠ [-1, -1, 2]
 * ]
 * 
 * 
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        LinkedList<List<Integer>> ans = new LinkedList<>();
        if(nums.length<3)
        {
            return ans;
        }

        Arrays.sort(nums);
        int m ,r ;
        for(int l=0;l<nums.length;++l)
        {
            if(nums[l]>0) break;
            if(l>0&&nums[l]==nums[l-1]) continue;
            m = l+1;
            r  = nums.length-1;
            while(m<r)
            {
                if(m>l+1&&nums[m]==nums[m-1])
                {
                    ++m;
                }else if(r<nums.length-1&&nums[r]==nums[r+1])
                {
                    --r;
                }else{
                    int d = nums[l]+nums[m]+nums[r];
                    if(d>0)
                    {
                        --r;
                    }else if(d<0)
                    {
                        ++m;
                    }else{
                        LinkedList<Integer> res = new LinkedList<>();
                        res.add(nums[l]);
                        res.add(nums[m++]);
                        res.add(nums[r--]);
                        ans.add(res);
                    }
                }
            }
        }
        return (List)ans;



    }
}
// @lc code=end

File: code/15.三数之和.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem15.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end() );
        int n = nums.size();
        
        vector< vector<int >> res;
        for(int i=0;i<n;i++) {
            if(nums[i] > 0) break;
            if (i>0 &&nums[i] == nums[i-1]) continue;
            int a= nums[i];

            
            int l = i+1,r = n-1;
            while(l <r ) {
                int b = nums[l],c = nums[r];
                if (a+b+c == 0) {
                    res.push_back({a,b,c});
                    while(l<r && nums[l] == nums[l+1]) ++l;
                    while(l<r && nums[r] == nums[r-1]) --r;
                    
                    
                    ++l;
                }else if (a+b+c>0) {

                    --r;

                }else {
                    //<0
                    ++l;
                }
            }
        }
        return res;
    }
};
// @lc code=end
